tiledb/api/src/query/conditions.rs

```diff
@@ -0,0 +1,754 @@
+cmp_value_impl!(i64, CmpValue::Int64);
+cmp_value_impl!(f32, CmpValue::Float32);
+cmp_value_impl!(f64, CmpValue::Float64);
+cmp_value_impl!(String, CmpValue::String);
+
+impl From<&str> for CmpValue {
+    fn from(val: &str) -> CmpValue {
+        CmpValue::String(val.to_string())
+    }
+}
+
+// N.B. I initially tried slices here, but that breaks the Deserialize trait.
+#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
+pub enum SetMemberValue {
+    UInt8(Vec<u8>),
+    UInt16(Vec<u16>),
+    UInt32(Vec<u32>),
+    UInt64(Vec<u64>),
+    Int8(Vec<i8>),
+    Int16(Vec<i16>),
+    Int32(Vec<i32>),
+    Int64(Vec<i64>),
+    Float32(Vec<f32>),
+    Float64(Vec<f64>),
+    String(Vec<String>),
+}
+
+macro_rules! slice_to_ptr_and_size {
+    ($val:expr) => {
+        (
+            $val.as_ptr() as *const std::ffi::c_void,
+            std::mem::size_of_val($val) as u64,
+        )
+    };
+}
+
+impl SetMemberValue {
+    fn len(&self) -> usize {
+        match self {
+            Self::UInt8(val) => val.len(),
+            Self::UInt16(val) => val.len(),
+            Self::UInt32(val) => val.len(),
+            Self::UInt64(val) => val.len(),
+            Self::Int8(val) => val.len(),
+            Self::Int16(val) => val.len(),
+            Self::Int32(val) => val.len(),
+            Self::Int64(val) => val.len(),
+            Self::Float32(val) => val.len(),
+            Self::Float64(val) => val.len(),
+            Self::String(val) => val.len(),
+        }
+    }
+
+    fn elem_size(&self) -> usize {
+        match self {
+            Self::UInt8(_) => std::mem::size_of::<u8>(),
+            Self::UInt16(_) => std::mem::size_of::<u16>(),
+            Self::UInt32(_) => std::mem::size_of::<u32>(),
+            Self::UInt64(_) => std::mem::size_of::<u64>(),
+            Self::Int8(_) => std::mem::size_of::<i8>(),
+            Self::Int16(_) => std::mem::size_of::<i16>(),
+            Self::Int32(_) => std::mem::size_of::<i32>(),
+            Self::Int64(_) => std::mem::size_of::<i64>(),
+            Self::Float32(_) => std::mem::size_of::<f32>(),
+            Self::Float64(_) => std::mem::size_of::<f64>(),
+            Self::String(_) => 0,
+        }
+    }
+
+    fn as_ptr_and_size(&self) -> (*const std::ffi::c_void, u64) {
+        match self {
+            Self::UInt8(val) => slice_to_ptr_and_size!(val),
+            Self::UInt16(val) => slice_to_ptr_and_size!(val),
+            Self::UInt32(val) => slice_to_ptr_and_size!(val),
+            Self::UInt64(val) => slice_to_ptr_and_size!(val),
+            Self::Int8(val) => slice_to_ptr_and_size!(val),
+            Self::Int16(val) => slice_to_ptr_and_size!(val),
+            Self::Int32(val) => slice_to_ptr_and_size!(val),
+            Self::Int64(val) => slice_to_ptr_and_size!(val),
+            Self::Float32(val) => slice_to_ptr_and_size!(val),
+            Self::Float64(val) => slice_to_ptr_and_size!(val),
+            Self::String(_) => (std::ptr::null(), 0),
```

Comment:
I don't think going down to `Vec<Vec<u8>>` is correct here based on the QueryCondition implementation constraints. Though I see your point and the Option is definitely cleaner and more Rust-y so I'll make that switch. In hindsight this seems like my brain pulling in C++'isms rather than updating to Rust Option result style.